### ISSUE TITLE
helm v2.12.3 return `.Release.IsUpgrade` false for `helm rollback` operation

### ISSUE THREAD

**Issue by AnitaNayak** [helm/helm#6088]
helm v2.12.3 return `.Release.IsUpgrade` false for `helm rollback` operation.
As per doc `.Release.IsUpgrade` should return true for both `helm upgrade/rollback` operations.
Its returning true for `helm upgrade` operation. But returns false for rollback.

```
[ test]# kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.5+icp-ee", GitCommit:"eb4df6c6fb47f5b4fd1ed8bfbfe2d0ed5ea636e1", GitTreeState:"clean", BuildDate:"2019-05-08T02:18:32Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/ppc64le"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.5+icp-ee", GitCommit:"eb4df6c6fb47f5b4fd1ed8bfbfe2d0ed5ea636e1", GitTreeState:"clean", BuildDate:"2019-05-08T02:18:32Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/ppc64le"}
[ test]#
[test]# helm version --tls
Client: &version.Version{SemVer:"v2.12.3", GitCommit:"eecf22f77df5f65c823aacd2dbd30ae6c65f186e", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.12.3+icp", GitCommit:"34e12adfe271fd157db8f9745affe84c0f603809", GitTreeState:"clean"}
[ test]#
```
Cloud Provider/Platform (ICP): 
Is there any other environment value or any other factor that affects this helm property?

**Comment by hickeyma** [comment]
@AnitaNayak You seem to be using a non community version of Helm. Could you try it with latest version of Helm and see if the issue still occurs?

**Comment by nweathe** [comment]
@hickeyma The non-community version of Helm is built on top of the community `v2.12.3` version of Helm. Can you confirm that the `.Release.IsUpgrade` functionality in the community `v2.12.3` version of Helm matches the documentation [here](https://github.com/helm/helm/blob/master/docs/chart_template_guide/builtin_objects.md) (i.e. `helm rollback` results in `.Release.IsUpgrade=true`)?

Thank you for your help!

**Comment by hickeyma** [comment]
Looks like a bug alright. Good catch, thanks. Tested as follows:

```console
$ cat chrt-6088/templates/configmap.yaml 
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-configmap
data:
  myvalue: "Hello World"
  {{- if .Release.IsUpgrade }}
  duck: "Upgrade is true"
  {{- else }}
  duck: "Upgrade is false"
  {{- end -}}

$ helm install --name chrt-6088 chrt-6088/
NAME:   chrt-6088
LAST DEPLOYED: Fri Oct  4 16:27:30 2019
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/ConfigMap
NAME                 DATA  AGE
chrt-6088-configmap  2     0s

$ kubectl get configmap chrt-6088-configmap -o yaml
apiVersion: v1
data:
  duck: Upgrade is false
  myvalue: Hello World
kind: ConfigMap
metadata:
  creationTimestamp: "2019-10-04T15:27:30Z"
  name: chrt-6088-configmap
  namespace: default
  resourceVersion: "2044"
  selfLink: /api/v1/namespaces/default/configmaps/chrt-6088-configmap
  uid: 7ab477cc-e6bb-11e9-9737-6ab98828a89d

$ helm upgrade chrt-6088 chrt-6088/
Release "chrt-6088" has been upgraded.
LAST DEPLOYED: Fri Oct  4 16:28:34 2019
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/ConfigMap
NAME                 DATA  AGE
chrt-6088-configmap  2     64s

$ kubectl get configmap chrt-6088-configmap -o yaml
apiVersion: v1
data:
  duck: Upgrade is true
  myvalue: Hello World
kind: ConfigMap
metadata:
  creationTimestamp: "2019-10-04T15:27:30Z"
  name: chrt-6088-configmap
  namespace: default
  resourceVersion: "2157"
  selfLink: /api/v1/namespaces/default/configmaps/chrt-6088-configmap
  uid: 7ab477cc-e6bb-11e9-9737-6ab98828a89d

$ helm rollback chrt-6088 1
Rollback was a success.
usr1@helm-v2:~/test/helm-charts$ helm history chrt-6088
REVISION	UPDATED                 	STATUS    	CHART          	APP VERSION	DESCRIPTION     
1       	Fri Oct  4 16:27:30 2019	SUPERSEDED	chrt-6088-0.1.0	1.0        	Install complete
2       	Fri Oct  4 16:28:34 2019	SUPERSEDED	chrt-6088-0.1.0	1.0        	Upgrade complete
3       	Fri Oct  4 16:29:30 2019	DEPLOYED  	chrt-6088-0.1.0	1.0        	Rollback to 1   

$ kubectl get configmap chrt-6088-configmap -o yaml
apiVersion: v1
data:
  duck: Upgrade is false
  myvalue: Hello World
kind: ConfigMap
metadata:
  creationTimestamp: "2019-10-04T15:27:30Z"
  name: chrt-6088-configmap
  namespace: default
  resourceVersion: "2254"
  selfLink: /api/v1/namespaces/default/configmaps/chrt-6088-configmap
  uid: 7ab477cc-e6bb-11e9-9737-6ab98828a89d
```

**Comment by bacongobbler** [comment]
Gentle reminder that the support window for bugfixes closes in a month. If nobody works on this, we'll have to close this out and consider this as Helm 2's intended behaviour.

**Comment by jeff-knurek** [comment]
hmmm, this is interesting. the only time that `.Release.IsUpgrade` is `false` is when you rollback to revision 1 OR another revision that was a rollback of revision 1. Meaning, the value of `.Release.IsUpgrade` is determined based on what the value was of that release.
so in @hickeyma example above:
```bash
$ helm rollback chrt-6088 1
Rollback was a success.
usr1@helm-v2:~/test/helm-charts$ helm history chrt-6088
REVISION	UPDATED                 	STATUS    	CHART          	APP VERSION	DESCRIPTION     
1       	Fri Oct  4 16:27:30 2019	SUPERSEDED	chrt-6088-0.1.0	1.0        	Install complete
2       	Fri Oct  4 16:28:34 2019	SUPERSEDED	chrt-6088-0.1.0	1.0        	Upgrade complete
3       	Fri Oct  4 16:29:30 2019	DEPLOYED  	chrt-6088-0.1.0	1.0        	Rollback to 1   
```
if he now does `helm rollback chrt-6088 2`, then `.Release.IsUpgrade` will be `true`

How to address this could impact other parts of the way the Release values are managed with rollback. For example, `Release.Time` (aka: `The time of the release`) during a rollback is also the original release's time, and not the time the rollback happened. I guess both of these should be addressed accordingly, because `Release.Revision` does increment during rollback. 
I'll try to create a PR that updates `IsUpgrade` & `Time`, but would like to have more discussion on what the desired behavior _should_ be.

**Comment by jeff-knurek** [comment]
I looked a bit more into the code, and saw that how the rollback works is that it takes the previous manifests, and applies them in a new release. and I think the idea of rollback is not to manipulate the manifests in any way _(I can't remember the term for that)_. for example in the ConfigMap used above, where the value of `duck` is set based on what the value of `.Release.IsUpgrade` is at the time _(revision 1: false, revision 2: true, revision 3: false, ...)_, if during the deploy of revision 3 the value of `duck` was changed, then it wouldn't really be revision 1 installed, but rather a new, and potentially unknown state, installed

----
in summary, I don't think this is a bug in the code, but rather poor documentation. 
> and i've already struggled in two long posts to describe the workings, so I'm not about to attempt to fix the documentation :)

**Comment by hickeyma** [comment]
Great investigation on this @jeff-knurek. Would you be willing to push PRs for the v2 and v3 docs?
